src/Conversion/ONNXToKrnl/Tensor/NonZero.cpp

```diff
@@ -0,0 +1,138 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===------------------- NonZero.cpp - Lowering NonZero Op ----------------===//
+//
+// Copyright 2019 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX NonZero Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/ONNX/ONNXShapeHelper.hpp"
+
+using namespace mlir;
+
+struct ONNXNonZeroOpLowering : public ConversionPattern {
+  ONNXNonZeroOpLowering(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXNonZeroOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXNonZeroOpAdaptor operandAdaptor(operands);
+    auto loc = op->getLoc();
+
+    // Builder helper.
+    KrnlBuilder createKrnl(rewriter, loc);
+    IndexExprScope outerScope(rewriter, loc);
+
+    // Common information.
+    Value X = operandAdaptor.X();
+    MemRefType xMemRefType = X.getType().cast<MemRefType>();
+    MemRefType resMemRefType = convertToMemRefType(*op->result_type_begin());
+    Type xElementType = xMemRefType.getElementType();
+    Type resElementType = resMemRefType.getElementType();
+    int64_t xRank = xMemRefType.getRank();
+
+    // Constant values.
+    Value iZero = emitConstantOp(rewriter, loc, rewriter.getIndexType(), 0);
+    Value iOne = emitConstantOp(rewriter, loc, rewriter.getIndexType(), 1);
+    Value zero = emitConstantOp(rewriter, loc, xElementType, 0);
+    LiteralIndexExpr litZero = LiteralIndexExpr(0);
```

Comment:
Nit: but this can simply be written as
```
LiteralIndexExpr litZero(0)
LiteralIndexExpr litOne(1)
```
 Optionally, iZero is also litZero.getValue() since they are of type index... but your code for that one is probably better.

Reply:
Yes, `LiteralIndexExpr litZero(0)` is definitely better.